From 8bd83b68cbc2815b0cc762d9134b4394b22b10e7 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Tue, 4 Nov 2008 16:47:09 +0200
Subject: [PATCH 01/17] interpol-test: remove unused include getopt.h

---
 src/tests/interpol-test.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/src/tests/interpol-test.c b/src/tests/interpol-test.c
index dd24e82..de13b8e 100644
--- a/src/tests/interpol-test.c
+++ b/src/tests/interpol-test.c
@@ -28,7 +28,6 @@
 #include <assert.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <getopt.h>
 #include <math.h>
 
 #include <pulse/pulseaudio.h>
-- 
1.6.2.rc1.13.gfd76c.dirty

