From 2d86a846d6233617d52d90fe5d7a336b58486251 Mon Sep 17 00:00:00 2001
From: Oliver Grawert <ogra@ubuntu.com>
Date: Sun, 18 Nov 2007 20:36:17 +0100
Subject: [PATCH] Double esound maximum sample size

To make sure the login and logout gnome sounds can be processed on remote
connections.
---
 src/pulsecore/protocol-esound.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/pulsecore/protocol-esound.c b/src/pulsecore/protocol-esound.c
index 70be848..e59c7da 100644
--- a/src/pulsecore/protocol-esound.c
+++ b/src/pulsecore/protocol-esound.c
@@ -72,7 +72,7 @@
 #define RECORD_BUFFER_SECONDS (5)
 #define RECORD_BUFFER_FRAGMENTS (100)
 
-#define MAX_CACHE_SAMPLE_SIZE (1024000)
+#define MAX_CACHE_SAMPLE_SIZE (2048000)
 
 #define SCACHE_PREFIX "esound."
 
