From f88121dc0bfe43d8f506b2b7a8cbbad046fd0112 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 20 Nov 2008 16:59:13 +0200
Subject: [PATCH 03/24] stream-restore: post a module change in trigger_save

---
 src/modules/module-stream-restore.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index d935caf..9d92b0d 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -229,6 +229,8 @@ static void trigger_save(struct userdata *u) {
         pa_pstream_send_tagstruct(pa_native_connection_get_pstream(c), t);
     }
 
+    pa_subscription_post(u->core, PA_SUBSCRIPTION_EVENT_MODULE|PA_SUBSCRIPTION_EVENT_CHANGE, u->module->index);
+
     if (u->save_time_event)
         return;
 
-- 
1.6.2.rc1.13.g25487f

