From 8f729b6959d2ddd45f0545d4f3fc0794aea4a0c7 Mon Sep 17 00:00:00 2001
From: Sjoerd Simons <sjoerd@luon.net>
Date: Sat, 10 May 2008 18:35:08 +0200
Subject: [PATCH] load module-gconf earlier

---
 src/daemon/default.pa.in |   18 +++++++++---------
 1 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/src/daemon/default.pa.in b/src/daemon/default.pa.in
index f6c68be..2dee693 100755
--- a/src/daemon/default.pa.in
+++ b/src/daemon/default.pa.in
@@ -64,6 +64,15 @@ load-module module-native-protocol-unix
 #load-module module-null-sink sink_name=rtp format=s16be channels=2 rate=44100 description="RTP Multicast Sink"
 #load-module module-rtp-send source=rtp.monitor
 
+### Load additional modules from GConf settings. This can be configured with the paprefs tool.
+### Please keep in mind that the modules configured by paprefs might conflict with manually
+### loaded modules.
+.ifexists module-gconf@PA_SOEXT@
+.nofail
+load-module module-gconf
+.fail
+.endif
+
 ### Automatically restore the volume of playback streams
 load-module module-volume-restore
 
@@ -95,15 +104,6 @@ load-module module-console-kit
 ### Enable positioned event sounds
 load-module module-position-event-sounds
 
-### Load additional modules from GConf settings. This can be configured with the paprefs tool.
-### Please keep in mind that the modules configured by paprefs might conflict with manually
-### loaded modules.
-.ifexists module-gconf@PA_SOEXT@
-.nofail
-load-module module-gconf
-.fail
-.endif
-
 ### Publish connection data in the X11 root window
 .ifexists module-x11-publish@PA_SOEXT@
 .nofail
