From c84aaa7e80dd5735e7f9565da47fc7326a0a787d Mon Sep 17 00:00:00 2001
From: CJ van den Berg <cj@vdbonline.com>
Date: Wed, 31 Oct 2007 13:59:31 +0100
Subject: [PATCH 03/17] Set ESD socket to /tmp/.esd/socket to match up with libesd

---
 src/daemon/default.pa.in |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/daemon/default.pa.in b/src/daemon/default.pa.in
index 57ccd9f..0c36c4d 100755
--- a/src/daemon/default.pa.in
+++ b/src/daemon/default.pa.in
@@ -64,7 +64,7 @@ load-module module-bluetooth-discover
 
 ### Load several protocols
 .ifexists module-esound-protocol-unix@PA_SOEXT@
-load-module module-esound-protocol-unix
+load-module module-esound-protocol-unix socket="/tmp/.esd/socket"
 .endif
 load-module module-native-protocol-unix
 
-- 
1.6.2.rc1.13.gfd76c.dirty

