From 2025e054009638ec11e5016fb0602ed0f5b330cf Mon Sep 17 00:00:00 2001
From: Luiz Augusto von Dentz <luiz.dentz@openbossa.org>
Date: Fri, 5 Dec 2008 15:48:48 -0300
Subject: [PATCH] Update bluetooth audio ipc header.

---
 src/modules/bluetooth/ipc.h |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/src/modules/bluetooth/ipc.h b/src/modules/bluetooth/ipc.h
index ae85e72..c785005 100644
--- a/src/modules/bluetooth/ipc.h
+++ b/src/modules/bluetooth/ipc.h
@@ -164,6 +164,8 @@ struct bt_getcapabilities_req {
 #define BT_MPEG_LAYER_2				(1 << 1)
 #define BT_MPEG_LAYER_3				1
 
+#define BT_SCO_NREC				1
+
 typedef struct {
 	uint8_t channel_mode;
 	uint8_t frequency;
@@ -183,12 +185,17 @@ typedef struct {
 	uint16_t bitrate;
 } __attribute__ ((packed)) mpeg_capabilities_t;
 
+typedef struct {
+	uint8_t features;
+	uint16_t sampling_rate;
+} __attribute__ ((packed)) sco_capabilities_t;
+
 struct bt_getcapabilities_rsp {
 	bt_audio_rsp_msg_header_t	rsp_h;
 	uint8_t				transport;	   /* Granted transport */
 	sbc_capabilities_t		sbc_capabilities;  /* A2DP only */
 	mpeg_capabilities_t		mpeg_capabilities; /* A2DP only */
-	uint16_t			sampling_rate;	   /* SCO only */
+	sco_capabilities_t		sco_capabilities;  /* SCO only */
 } __attribute__ ((packed));
 
 /* BT_SETCONFIGURATION_REQ */
-- 
1.6.0.2.514.g23abd3

