From 21337a97416e1df9e4890f478c905df092c7a947 Mon Sep 17 00:00:00 2001
From: Sarha Jyri (Nokia-D/Helsinki) <Jyri.Sarha@nokia.com>
Date: Wed, 1 Oct 2008 18:08:36 +0300
Subject: [PATCH] Fix supend

---
 src/pulsecore/sink.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/pulsecore/sink.c b/src/pulsecore/sink.c
index d8d1f79..6adab8a 100644
--- a/src/pulsecore/sink.c
+++ b/src/pulsecore/sink.c
@@ -556,7 +556,7 @@ static void inputs_drop(pa_sink *s, pa_mix_info *info, unsigned n, pa_memchunk *
         /* Drop read data */
         pa_sink_input_drop(i, result->length);
 
-        if (s->monitor_source && PA_SOURCE_IS_OPENED(pa_source_get_state(s->monitor_source))) {
+        if (s->monitor_source && PA_SOURCE_IS_OPENED(s->monitor_source->thread_info.state)) {
 
             if (pa_hashmap_size(i->thread_info.direct_outputs) > 0) {
                 void *ostate = NULL;
-- 
1.5.6.5

