From 78a4d1bcf68235fe8985c0f90e69612977e61dfc Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Wed, 12 Nov 2008 17:40:34 +0200
Subject: [PATCH] build: make sure we build libltdl before the rest

---
 Makefile.am  |    2 +-
 configure.ac |    1 +
 2 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 07c4271..01d4fd6 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -18,7 +18,7 @@
 ACLOCAL_AMFLAGS = -I m4
 
 EXTRA_DIST = bootstrap.sh LICENSE GPL LGPL doxygen/Makefile.am doxygen/Makefile.in doxygen/doxygen.conf.in README todo
-SUBDIRS=src doxygen man po
+SUBDIRS= libltdl src doxygen man po
 
 MAINTAINERCLEANFILES =
 noinst_DATA =
diff --git a/configure.ac b/configure.ac
index 2a04076..8ffc689 100644
--- a/configure.ac
+++ b/configure.ac
@@ -227,6 +227,7 @@ old_LIBS=$LIBS
 LIBS="$LIBS $LIBLTDL"
 AC_CHECK_FUNCS([lt_dlmutex_register])
 LIBS=$old_LIBS
+
 AC_CHECK_TYPES([struct lt_user_dlloader, lt_dladvise], , , [#include <ltdl.h>])
 
 if test "x$enable_ltdl_install" = "xno" && test "x$ac_cv_lib_ltdl_lt_dlinit" = "xno" ; then
-- 
1.6.0.2.514.g23abd3

