From 4db536d268fff5a6f129ed07884439a6b5170ab6 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 11 Dec 2008 22:05:31 +0200
Subject: [PATCH 10/26] build: install pulse tests under $(libexecdir)/pulse-tests (Nokia specific)

---
 src/Makefile.am |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index 540ebe1..6a9b0b4 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -29,6 +29,7 @@ pulsecoreincludedir=$(includedir)/pulsecore
 pulseconfdir=$(sysconfdir)/pulse
 pulselibexecdir=$(libexecdir)/pulse
 xdgautostartdir=$(sysconfdir)/xdg/autostart
+pulsetestdir=$(libexecdir)/pulse-tests
 
 ###################################
 #            Defines              #
@@ -255,7 +256,7 @@ TESTS = \
 		lock-autospawn-test \
 		prioq-test
 
-TESTS_BINARIES = \
+pulsetest_PROGRAMS = \
 		mainloop-test \
 		mcalign-test \
 		pacat-simple \
@@ -296,7 +297,7 @@ if HAVE_SIGXCPU
 #TESTS += \
 #		cpulimit-test \
 #		cpulimit-test2
-TESTS_BINARIES += \
+pulsetest_PROGRAMS += \
 		cpulimit-test \
 		cpulimit-test2
 endif
@@ -304,12 +305,12 @@ endif
 if HAVE_GLIB20
 TESTS += \
 		mainloop-test-glib
-TESTS_BINARIES += \
+pulsetest_PROGRAMS += \
 		mainloop-test-glib
 endif
 
 if HAVE_GTK20
-TESTS_BINARIES += \
+pulsetest_PROGRAMS += \
 		gtk-test
 endif
 
-- 
1.6.2.rc1.13.gfd76c.dirty

