From 8879b86b521308b049ac2b709f6b43aaf8e95669 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 4 Dec 2008 16:23:06 +0200
Subject: [PATCH] core: report remaining shared objects when cleanup

---
 src/pulsecore/shared.c |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/src/pulsecore/shared.c b/src/pulsecore/shared.c
index 77d919d..4c1ad80 100644
--- a/src/pulsecore/shared.c
+++ b/src/pulsecore/shared.c
@@ -111,7 +111,14 @@ void pa_shared_cleanup(pa_core *c) {
     if (!c->shared)
         return;
 
-    pa_assert(pa_hashmap_isempty(c->shared));
+    if (!pa_hashmap_isempty(c->shared)) {
+        pa_strbuf *s = pa_strbuf_new();
+
+        pa_shared_dump(c, s);
+        pa_log_debug(pa_strbuf_tostring(s));
+        pa_strbuf_free(s);
+        pa_assert(pa_hashmap_isempty(c->shared));
+    }
 
     pa_hashmap_free(c->shared, NULL, NULL);
     c->shared = NULL;
-- 
1.6.0.2.514.g23abd3

