From 589a87cb930b56416a6b2549709a5bd555b85642 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Tue, 10 Mar 2009 20:08:34 +0200
Subject: [PATCH 13/34] pulsecore: intermediate s-i soft_vol is 0dB

---
 src/pulsecore/sink-input.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/pulsecore/sink-input.c b/src/pulsecore/sink-input.c
index 6a18ff9..1f8826e 100644
--- a/src/pulsecore/sink-input.c
+++ b/src/pulsecore/sink-input.c
@@ -288,6 +288,8 @@ int pa_sink_input_new(
 
     i->volume_factor = data->volume_factor;
     pa_cvolume_init(&i->soft_volume);
+    if (i->origin_sink)
+      pa_cvolume_reset(&i->soft_volume, i->origin_sink->sample_spec.channels);
     i->save_volume = data->save_volume;
     i->save_sink = data->save_sink;
     i->save_muted = data->save_muted;
-- 
1.6.2.rc1.13.gfd76c.dirty

