From abbd3bf151a1607a714c45530bc70315e894d4f7 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Sat, 4 Apr 2009 14:50:19 +0300
Subject: [PATCH] stream-restore: update dump_db and add DEBUG_VOLUME

---
 src/modules/module-stream-restore.c |   12 +++++++-----
 1 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index 2c90e72..90d5cf6 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -592,8 +592,8 @@ static void apply_entry(struct userdata *u, const char *name, struct entry *e) {
     }
 }
 
-#if 0
-static void dump_database(struct userdata *u) {
+#ifdef DEBUG_VOLUME
+PA_GCC_UNUSED static void stream_restore_dump_database(struct userdata *u) {
     datum key;
 
     key = gdbm_firstkey(u->gdbm_file);
@@ -610,10 +610,12 @@ static void dump_database(struct userdata *u) {
         if ((e = read_entry(u, name))) {
             char t[256];
             pa_log("name=%s", name);
-            pa_log("device=%s", e->device);
+            pa_log("version=%d", e->version);
+            pa_log("relative volume=%s", e->relative_volume_valid ? pa_cvolume_snprint(t, sizeof(t), &e->relative_volume) : "");
+            pa_log("absolute volume=%s", e->absolute_volume_valid ? pa_cvolume_snprint(t, sizeof(t), &e->absolute_volume) : "");
+            pa_log("mute=%s", e->muted_valid ? pa_yes_no(e->muted) : "");
             pa_log("channel_map=%s", pa_channel_map_snprint(t, sizeof(t), &e->channel_map));
-            pa_log("volume=%s", pa_cvolume_snprint(t, sizeof(t), &e->volume));
-            pa_log("mute=%s", pa_yes_no(e->muted));
+            pa_log("device=%s", e->device_valid ? e->device : "");
             pa_xfree(e);
         }
 
-- 
1.5.6.3

