From 2359b8eab7e483bcc01d3da6cd3d5632f82acf82 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 27 Nov 2008 16:14:00 +0200
Subject: [PATCH] stream-restore: add module usage

---
 src/modules/module-stream-restore.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index 8cb3a89..028a299 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -58,6 +58,9 @@ PA_MODULE_AUTHOR("Lennart Poettering");
 PA_MODULE_DESCRIPTION("Automatically restore the volume/mute/device state of streams");
 PA_MODULE_VERSION(PACKAGE_VERSION);
 PA_MODULE_LOAD_ONCE(TRUE);
+PA_MODULE_USAGE("restore_device=<boolean> "
+		"restore_volume=<boolean> "
+		"restore_muted=<boolean>");
 
 #define SAVE_INTERVAL 10
 
-- 
1.6.0.2.514.g23abd3

