From c37feab0555ef743f8d9944b41159747c08b84c9 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Tue, 10 Mar 2009 17:55:37 +0200
Subject: [PATCH 20/32] pulsecore: fix a comment about pa_sink_set_volume

---
 src/pulsecore/sink.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/pulsecore/sink.c b/src/pulsecore/sink.c
index 87e3374..1544f7a 100644
--- a/src/pulsecore/sink.c
+++ b/src/pulsecore/sink.c
@@ -1015,7 +1015,7 @@ static void update_input_soft_volume(pa_sink *s, const pa_cvolume *new_volume, c
         /* We don't issue PA_SINK_INPUT_MESSAGE_SET_VOLUME because
          * we want the update to have atomically with the sink
          * volume update, hence we do it within the
-         * pa_sink_set_flat_volume() call below*/
+         * pa_sink_set_volume() call below*/
     }
 }
 
@@ -1031,7 +1031,7 @@ void pa_sink_update_flat_volume(pa_sink *s, pa_cvolume *new_volume, pa_sink_inpu
      * might need to fix up the sink volume accordingly. Please note
      * that we don't actually update the sinks volume here, we only
      * return how it needs to be updated. The caller should then call
-     * pa_sink_set_flat_volume().*/
+     * pa_sink_set_volume().*/
 
     if (pa_idxset_isempty(s->inputs)) {
         /* In the special case that we have no sink input we leave the
-- 
1.6.2.rc1.13.gfd76c.dirty

