From 00cd1d6495f390e50cff2e378dc25f4971d05cfe Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Sat, 4 Apr 2009 14:47:30 +0300
Subject: [PATCH 21/68] cli-text: print internal volume for debugging purposes

---
 src/pulsecore/cli-text.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/src/pulsecore/cli-text.c b/src/pulsecore/cli-text.c
index 604678b..7712f9e 100644
--- a/src/pulsecore/cli-text.c
+++ b/src/pulsecore/cli-text.c
@@ -235,6 +235,7 @@ char *pa_sink_list_to_string(pa_core *c) {
             "\tvolume: %s%s%s\n"
             "\t        balance %0.2f\n"
             "\tbase volume: %s%s%s\n"
+            "\tsoft volume: %s%s%s\n"
             "\tvolume steps: %u\n"
             "\tmuted: %s\n"
             "\tcurrent latency: %0.2f ms\n"
@@ -265,6 +266,9 @@ char *pa_sink_list_to_string(pa_core *c) {
             pa_volume_snprint(v, sizeof(v), sink->base_volume),
             sink->flags & PA_SINK_DECIBEL_VOLUME ? "\n\t             " : "",
             sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_volume_snprint_dB(vdb, sizeof(vdb), sink->base_volume) : "",
+            pa_cvolume_snprint(v, sizeof(v), &sink->soft_volume),
+            sink->flags & PA_SINK_DECIBEL_VOLUME ? "\n\t             " : "",
+            sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_cvolume_snprint_dB(vdb, sizeof(vdb), &sink->soft_volume) : "",
             sink->n_volume_steps,
             pa_yes_no(pa_sink_get_mute(sink, FALSE)),
             (double) pa_sink_get_latency(sink) / (double) PA_USEC_PER_MSEC,
@@ -530,6 +534,8 @@ char *pa_sink_input_list_to_string(pa_core *c) {
             "\tvolume: %s\n"
             "\t        %s\n"
             "\t        balance %0.2f\n"
+            "\tsoft volume: %s\n"
+            "\tvolume factor: %s\n"
             "\tmuted: %s\n"
             "\tcurrent latency: %0.2f ms\n"
             "\trequested latency: %s\n"
@@ -553,6 +559,8 @@ char *pa_sink_input_list_to_string(pa_core *c) {
             pa_cvolume_snprint(cv, sizeof(cv), &v),
             pa_sw_cvolume_snprint_dB(cvdb, sizeof(cvdb), &v),
             pa_cvolume_get_balance(&v, &i->channel_map),
+            pa_cvolume_snprint(cv, sizeof(cv), &i->soft_volume),
+            pa_cvolume_snprint(cv, sizeof(cv), &i->volume_factor),
             pa_yes_no(pa_sink_input_get_mute(i)),
             (double) pa_sink_input_get_latency(i, NULL) / PA_USEC_PER_MSEC,
             clt,
-- 
1.6.3.1

