From 8ca765edf0570c1e3a6b58aa35b45aeae33b23b1 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 11 Dec 2008 22:05:31 +0200
Subject: [PATCH] build: install pulse tests under $(libexecdir)/pulse-tests (Nokia specific)

---
 src/Makefile.am |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index a2a6b64..51a3e1c 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -29,6 +29,7 @@ pulsecoreincludedir=$(includedir)/pulsecore
 pulseconfdir=$(sysconfdir)/pulse
 pulselibexecdir=$(libexecdir)/pulse
 xdgautostartdir=$(sysconfdir)/xdg/autostart
+pulsetestdir=$(libexecdir)/pulse-tests
 
 ###################################
 #            Defines              #
@@ -248,7 +249,7 @@ TESTS = \
 		lock-autospawn-test \
 		prioq-test
 
-noinst_PROGRAMS = \
+pulsetest_PROGRAMS = \
 		mainloop-test \
 		mcalign-test \
 		pacat-simple \
@@ -288,7 +289,7 @@ if HAVE_SIGXCPU
 #TESTS += \
 #		cpulimit-test \
 #		cpulimit-test2
-noinst_PROGRAMS += \
+pulsetest_PROGRAMS += \
 		cpulimit-test \
 		cpulimit-test2
 endif
@@ -296,7 +297,7 @@ endif
 if HAVE_GLIB20
 TESTS += \
 		mainloop-test-glib
-noinst_PROGRAMS += \
+pulsetest_PROGRAMS += \
 		mainloop-test-glib
 endif
 
-- 
1.6.0.2.514.g23abd3

