From a1c0725b737b9cf1433cd5ecfae6967ac71e9a6b Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Mon, 2 Mar 2009 18:52:36 +0200
Subject: [PATCH 28/30] pulsecore: hack a test for loop usage

---
 src/pulse/mainloop.c |   23 ++++++++++++++++++++++-
 1 files changed, 22 insertions(+), 1 deletions(-)

diff --git a/src/pulse/mainloop.c b/src/pulse/mainloop.c
index 60e5d1f..3403a6f 100644
--- a/src/pulse/mainloop.c
+++ b/src/pulse/mainloop.c
@@ -52,6 +52,7 @@
 #include <pulsecore/core-error.h>
 #include <pulsecore/winsock.h>
 #include <pulsecore/macro.h>
+#include <pulsecore/once.h>
 
 #include "mainloop.h"
 
@@ -853,7 +854,27 @@ int pa_mainloop_poll(pa_mainloop *m) {
     else {
         pa_assert(!m->rebuild_pollfds);
 
-        if (m->poll_func)
+#if 1
+	{
+            static struct timeval start;
+            static struct timeval now;
+            static int hz = 0;
+
+            ++hz;
+
+            PA_ONCE_BEGIN {
+                pa_gettimeofday(&start);
+            } PA_ONCE_END;
+
+            if (pa_timeval_diff(pa_gettimeofday(&now), &start) >= PA_USEC_PER_SEC) {
+                pa_log_debug("Before poll(): %d nfds, %d Hz", m->n_pollfds, hz);
+                hz = 0;
+                pa_gettimeofday(&start);
+            }
+	}
+#endif
+
+	if (m->poll_func)
             m->poll_func_ret = m->poll_func(m->pollfds, m->n_pollfds, m->prepared_timeout, m->poll_func_userdata);
         else
             m->poll_func_ret = poll(m->pollfds, m->n_pollfds, m->prepared_timeout);
-- 
1.6.2.rc1.13.gfd76c.dirty

