From e83621724d7939b97b4f01f0d7e965d61ef8e55e Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Thu, 19 Mar 2009 16:48:31 +0200
Subject: [PATCH 34/37] bluetooth: rework the info_valid logic to be more tolerant

---
 src/modules/bluetooth/bluetooth-util.c |    5 ++---
 1 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/src/modules/bluetooth/bluetooth-util.c b/src/modules/bluetooth/bluetooth-util.c
index 5213ad6..912697e 100644
--- a/src/modules/bluetooth/bluetooth-util.c
+++ b/src/modules/bluetooth/bluetooth-util.c
@@ -114,15 +114,14 @@ void pa_bluetooth_device_free(pa_bluetooth_device *d) {
 static pa_bool_t device_is_loaded(pa_bluetooth_device *d) {
     pa_assert(d);
 
-    return d->device_info_valid && d->audio_sink_info_valid && d->headset_info_valid;
+    return d->device_info_valid && (d->audio_sink_info_valid || d->headset_info_valid);
 }
 
 static pa_bool_t device_is_audio(pa_bluetooth_device *d) {
     pa_assert(d);
 
     pa_assert(d->device_info_valid);
-    pa_assert(d->audio_sink_info_valid);
-    pa_assert(d->headset_info_valid);
+    pa_assert(d->audio_sink_info_valid || d->headset_info_valid);
 
     return d->device_info_valid > 0 &&
         (d->audio_sink_info_valid > 0 || d->headset_info_valid > 0);
-- 
1.6.2.rc1.13.gfd76c.dirty

