From 5df4bea84d4b7b1edde874d427c7f55341c74e1e Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Thu, 11 Dec 2008 22:05:31 +0200
Subject: [PATCH] build: install pulse tests under $(libexecdir)/pulse-tests

---
 src/Makefile.am |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index ffa276d..a6302ac 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -29,6 +29,7 @@ pulsecoreincludedir=$(includedir)/pulsecore
 pulseconfdir=$(sysconfdir)/pulse
 pulselibexecdir=$(libexecdir)/pulse
 xdgautostartdir=$(sysconfdir)/xdg/autostart
+pulsetestdir=$(libexecdir)/pulse-tests
 
 ###################################
 #            Defines              #
@@ -235,7 +236,7 @@ pabrowse_LDFLAGS = $(AM_LDFLAGS) $(BINLDFLAGS)
 #         Test programs           #
 ###################################
 
-noinst_PROGRAMS = \
+pulsetest_PROGRAMS = \
 		mainloop-test \
 		mcalign-test \
 		pacat-simple \
@@ -272,13 +273,13 @@ noinst_PROGRAMS = \
 		prioq-test
 
 if HAVE_SIGXCPU
-noinst_PROGRAMS += \
+pulsetest_PROGRAMS += \
 		cpulimit-test \
 		cpulimit-test2
 endif
 
 if HAVE_GLIB20
-noinst_PROGRAMS += \
+pulsetest_PROGRAMS += \
 		mainloop-test-glib
 endif
 
-- 
1.6.0.2.514.g23abd3

