From 21c566c10998410c98275595f9324bd33406b262 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marcandre.lureau@gmail.com>
Date: Fri, 9 Jan 2009 16:59:05 +0200
Subject: [PATCH 43/66] build: add pulsecore install-hook (Nokia specific)

---
 src/Makefile.am |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/src/Makefile.am b/src/Makefile.am
index 7962a37..d49f4fd 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1619,11 +1619,18 @@ install-exec-hook:
 	ln -sf pacat $(DESTDIR)$(bindir)/parec
 	rm -f $(DESTDIR)$(libdir)/libpulsedsp.la
 	rm -f $(DESTDIR)$(modlibexecdir)/*.la
+# Nokia pulsecore hack
+	mkdir -p $(DESTDIR)$(includedir)/pulsecore/modules
+	cp $(top_builddir)/config.h $(DESTDIR)$(includedir)/pulsecore/pulsecore-config.h
+	cp pulsecore/*.h $(DESTDIR)$(includedir)/pulsecore
+	cp modules/*-util.h $(DESTDIR)$(includedir)/pulsecore/modules
 
 uninstall-hook:
 	rm -f $(DESTDIR)$(bindir)/parec
 	rm -f $(DESTDIR)$(libdir)/libpulsedsp.*
 	rm -f $(DESTDIR)$(modlibexecdir)/*.so
+# Nokia pulsecore hack
+	rm -rf $(DESTDIR)$(includedir)/pulsecore
 
 massif: pulseaudio
 	libtool --mode=execute valgrind --tool=massif --depth=6  --alloc-fn=pa_xmalloc --alloc-fn=pa_xmalloc0 --alloc-fn=pa_xrealloc --alloc-fn=dbus_realloc --alloc-fn=pa_xnew0_internal --alloc-fn=pa_xnew_internal ./pulseaudio
-- 
1.6.3.1

