From 179564f8357cd5e2220f462a42eb158967d73379 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Mon, 1 Jun 2009 14:17:59 +0300
Subject: [PATCH 56/70] mainloop: log as deprecated timeval time_new

---
 src/pulse/mainloop.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/src/pulse/mainloop.c b/src/pulse/mainloop.c
index 9b4fc9f..f0dd0a2 100644
--- a/src/pulse/mainloop.c
+++ b/src/pulse/mainloop.c
@@ -413,6 +413,8 @@ static pa_time_event* mainloop_time_new(
         void *userdata) {
     pa_usec_t t;
 
+    pa_log("*deprecated* new timeval time event!");
+
     if (tv) {
         struct timeval now;
 
-- 
1.6.3.1

