From 89397b1024f173c6bde6818f7ebec9c409c60e8d Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Tue, 23 Jun 2009 19:38:55 +0300
Subject: [PATCH 63/75] stream-restore: don't override volume entry, when applied to different channel maps

---
 src/modules/module-stream-restore.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index f5cbbb5..125f1e5 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -788,8 +788,11 @@ static void apply_entry(struct userdata *u, const char *name, struct entry *e) {
         pa_xfree(n);
 
         if (u->restore_volume && e->volume_valid) {
-            pa_log_info("Restoring volume for sink input %s.", name);
-            pa_sink_input_set_volume(si, pa_cvolume_remap(&e->volume, &e->channel_map, &si->channel_map), TRUE, e->volume_is_absolute);
+            pa_cvolume v;
+
+            pa_log_info("Restoring volume for sink input %s. c %d vol %d", name, e->channel_map.channels, e->volume.values[0]);
+            v = e->volume;
+            pa_sink_input_set_volume(si, pa_cvolume_remap(&v, &e->channel_map, &si->channel_map), TRUE, e->volume_is_absolute);
         }
 
         if (u->restore_muted && e->muted_valid) {
-- 
1.6.3.3

