From 709c5902d1d963f870f63b4527483c6f26e21a12 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Tue, 23 Jun 2009 15:52:37 +0300
Subject: [PATCH 65/66] stream-restore: workaroung moving stream volume restore

---
 src/modules/module-stream-restore.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index 761b293..972b8a5 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -776,7 +776,10 @@ static void apply_entry(struct userdata *u, const char *name, struct entry *e) {
         char *n;
         pa_sink *s;
 
-        if (!(n = get_name(si->proplist, "sink-input")))
+        if (!si->sink) /* for eg. moving */
+            continue;
+
+	if (!(n = get_name(si->proplist, "sink-input")))
             continue;
 
         if (!pa_streq(name, n)) {
@@ -808,6 +811,9 @@ static void apply_entry(struct userdata *u, const char *name, struct entry *e) {
         char *n;
         pa_source *s;
 
+	if (!so->source) /* for eg. moving */
+            continue;
+
         if (!(n = get_name(so->proplist, "source-output")))
             continue;
 
-- 
1.6.3.1

