From edc24729f74f0b0a1f4ce3f26aba6d88d88de8cc Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Sat, 4 Apr 2009 14:47:30 +0300
Subject: [PATCH 69/77] cli-text: print internal volume for debugging purposes

---
 src/pulsecore/cli-text.c |   20 +++++++++++++++-----
 1 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/src/pulsecore/cli-text.c b/src/pulsecore/cli-text.c
index 604678b..86d8734 100644
--- a/src/pulsecore/cli-text.c
+++ b/src/pulsecore/cli-text.c
@@ -216,9 +216,11 @@ char *pa_sink_list_to_string(pa_core *c) {
     for (sink = pa_idxset_first(c->sinks, &idx); sink; sink = pa_idxset_next(c->sinks, &idx)) {
         char ss[PA_SAMPLE_SPEC_SNPRINT_MAX],
             cv[PA_CVOLUME_SNPRINT_MAX],
+            csv[PA_CVOLUME_SNPRINT_MAX],
             cvdb[PA_SW_CVOLUME_SNPRINT_DB_MAX],
-            v[PA_VOLUME_SNPRINT_MAX],
-            vdb[PA_SW_VOLUME_SNPRINT_DB_MAX],
+            bv[PA_VOLUME_SNPRINT_MAX],
+            svdb[PA_SW_VOLUME_SNPRINT_DB_MAX],
+            bvdb[PA_SW_VOLUME_SNPRINT_DB_MAX],
             cm[PA_CHANNEL_MAP_SNPRINT_MAX], *t;
         const char *cmn;
 
@@ -235,6 +237,7 @@ char *pa_sink_list_to_string(pa_core *c) {
             "\tvolume: %s%s%s\n"
             "\t        balance %0.2f\n"
             "\tbase volume: %s%s%s\n"
+            "\tsoft volume: %s%s%s\n"
             "\tvolume steps: %u\n"
             "\tmuted: %s\n"
             "\tcurrent latency: %0.2f ms\n"
@@ -262,9 +265,12 @@ char *pa_sink_list_to_string(pa_core *c) {
             sink->flags & PA_SINK_DECIBEL_VOLUME ? "\n\t        " : "",
             sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_cvolume_snprint_dB(cvdb, sizeof(cvdb), pa_sink_get_volume(sink, FALSE, FALSE)) : "",
             pa_cvolume_get_balance(pa_sink_get_volume(sink, FALSE, FALSE), &sink->channel_map),
-            pa_volume_snprint(v, sizeof(v), sink->base_volume),
+            pa_volume_snprint(bv, sizeof(bv), sink->base_volume),
+            sink->flags & PA_SINK_DECIBEL_VOLUME ? "\n\t             " : "",
+            sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_volume_snprint_dB(bvdb, sizeof(bvdb), sink->base_volume) : "",
+            pa_cvolume_snprint(csv, sizeof(csv), &sink->soft_volume),
             sink->flags & PA_SINK_DECIBEL_VOLUME ? "\n\t             " : "",
-            sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_volume_snprint_dB(vdb, sizeof(vdb), sink->base_volume) : "",
+            sink->flags & PA_SINK_DECIBEL_VOLUME ? pa_sw_cvolume_snprint_dB(svdb, sizeof(svdb), &sink->soft_volume) : "",
             sink->n_volume_steps,
             pa_yes_no(pa_sink_get_mute(sink, FALSE)),
             (double) pa_sink_get_latency(sink) / (double) PA_USEC_PER_MSEC,
@@ -504,7 +510,7 @@ char *pa_sink_input_list_to_string(pa_core *c) {
     pa_strbuf_printf(s, "%u sink input(s) available.\n", pa_idxset_size(c->sink_inputs));
 
     for (i = pa_idxset_first(c->sink_inputs, &idx); i; i = pa_idxset_next(c->sink_inputs, &idx)) {
-        char ss[PA_SAMPLE_SPEC_SNPRINT_MAX], cvdb[PA_SW_CVOLUME_SNPRINT_DB_MAX], cv[PA_CVOLUME_SNPRINT_MAX], cm[PA_CHANNEL_MAP_SNPRINT_MAX], *t, clt[28];
+        char ss[PA_SAMPLE_SPEC_SNPRINT_MAX], cvdb[PA_SW_CVOLUME_SNPRINT_DB_MAX], cv[PA_CVOLUME_SNPRINT_MAX], cm[PA_CHANNEL_MAP_SNPRINT_MAX], csv[PA_CVOLUME_SNPRINT_MAX], cvf[PA_CVOLUME_SNPRINT_MAX], *t, clt[28];
         pa_usec_t cl;
         const char *cmn;
         pa_cvolume v;
@@ -530,6 +536,8 @@ char *pa_sink_input_list_to_string(pa_core *c) {
             "\tvolume: %s\n"
             "\t        %s\n"
             "\t        balance %0.2f\n"
+            "\tsoft volume: %s\n"
+            "\tvolume factor: %s\n"
             "\tmuted: %s\n"
             "\tcurrent latency: %0.2f ms\n"
             "\trequested latency: %s\n"
@@ -553,6 +561,8 @@ char *pa_sink_input_list_to_string(pa_core *c) {
             pa_cvolume_snprint(cv, sizeof(cv), &v),
             pa_sw_cvolume_snprint_dB(cvdb, sizeof(cvdb), &v),
             pa_cvolume_get_balance(&v, &i->channel_map),
+            pa_cvolume_snprint(csv, sizeof(csv), &i->soft_volume),
+            pa_cvolume_snprint(cvf, sizeof(cvf), &i->volume_factor),
             pa_yes_no(pa_sink_input_get_mute(i)),
             (double) pa_sink_input_get_latency(i, NULL) / PA_USEC_PER_MSEC,
             clt,
-- 
1.6.3.3

