From 7d6f68979bc51c64954cac18e00eab302b8ae412 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Wed, 15 Jul 2009 20:21:22 +0300
Subject: [PATCH 74/84] db-simple: fix double free (to squash)

---
 src/pulsecore/database-simple.c |   27 ++++++++++++++-------------
 1 files changed, 14 insertions(+), 13 deletions(-)

diff --git a/src/pulsecore/database-simple.c b/src/pulsecore/database-simple.c
index 0fa7d1f..ceb565b 100644
--- a/src/pulsecore/database-simple.c
+++ b/src/pulsecore/database-simple.c
@@ -486,26 +486,27 @@ int pa_database_sync(pa_database *database) {
 
     f = fopen(db->tmp_filename, "w");
 
-    if (f) {
-        state = NULL;
-        while((e = pa_hashmap_iterate(db->map, &state, NULL))) {
-            if (write_entry(f, e)) {
-                pa_log_warn("error while writing to file. %s", pa_cstrerror(errno));
-                goto fail;
-            }
-        }
-
-        fclose(f);
+    if (!f)
+        goto fail;
 
-        if (rename(db->tmp_filename, db->filename) < 0) {
-            pa_log_warn("error while renaming file. %s", pa_cstrerror(errno));
+    state = NULL;
+    while((e = pa_hashmap_iterate(db->map, &state, NULL))) {
+        if (write_entry(f, e)) {
+            pa_log_warn("error while writing to file. %s", pa_cstrerror(errno));
             goto fail;
         }
-    } else {
+    }
+
+    fclose(f);
+    f = NULL;
+
+    if (rename(db->tmp_filename, db->filename) < 0) {
+        pa_log_warn("error while renaming file. %s", pa_cstrerror(errno));
         goto fail;
     }
 
     return 0;
+
 fail:
     if (f)
         fclose(f);
-- 
1.6.3.3

