From d57102969a6bfa761684003c339e26e375b41992 Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Marc-Andr=C3=A9=20Lureau?= <marc-andre.lureau@nokia.com>
Date: Mon, 20 Jul 2009 13:53:17 +0300
Subject: [PATCH 78/85] pulse: even in case of record stream, let's initialize req_bytes to 0

---
 src/pulse/stream.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/pulse/stream.c b/src/pulse/stream.c
index c9383a8..3ec1342 100644
--- a/src/pulse/stream.c
+++ b/src/pulse/stream.c
@@ -861,7 +861,7 @@ static void automatic_buffer_attr(pa_stream *s, pa_buffer_attr *attr, const pa_s
 
 void pa_create_stream_callback(pa_pdispatch *pd, uint32_t command, uint32_t tag, pa_tagstruct *t, void *userdata) {
     pa_stream *s = userdata;
-    uint32_t requested_bytes;
+    uint32_t requested_bytes = 0;
 
     pa_assert(pd);
     pa_assert(s);
-- 
1.6.3.3

