From 21aa19697c17207b4d06cc6ebac542a289691975 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Tue, 21 Apr 2009 21:23:53 +0200
Subject: [PATCH] llist: add PA_LLIST_FOREACH

---
 src/pulsecore/llist.h |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/src/pulsecore/llist.h b/src/pulsecore/llist.h
index 77a1749..58b51c6 100644
--- a/src/pulsecore/llist.h
+++ b/src/pulsecore/llist.h
@@ -104,4 +104,7 @@
         }                                                               \
     } while (0)
 
+#define PA_LLIST_FOREACH(i,head)                                        \
+    for (i = (head); i; i = i->next)
+
 #endif
-- 
1.6.0.4

